collect_app/src/main/java/org/odk/collect/android/formentry/FormEntryViewModel.java

```diff
@@ -0,0 +1,285 @@
+                        }
+                    }
+                }
+
+                setSaveRequestState(taskResult, SaveResult.State.SAVED, saveResult.request);
+                break;
+            }
+
+            case SaveToDiskTask.SAVE_ERROR: {
+                if (auditEventLogger != null) {
+                    auditEventLogger.logEvent(AuditEvent.AuditEventType.SAVE_ERROR, true, clock.getCurrentTime());
+                }
+
+                setSaveRequestState(taskResult, SaveResult.State.SAVE_ERROR, saveResult.request);
+                break;
+            }
+
+            case SaveToDiskTask.ENCRYPTION_ERROR: {
+                if (auditEventLogger != null) {
+                    auditEventLogger.logEvent(AuditEvent.AuditEventType.FINALIZE_ERROR, true, clock.getCurrentTime());
+                }
+
+                setSaveRequestState(taskResult, SaveResult.State.FINALIZE_ERROR, saveResult.request);
+                break;
+            }
+
+            case FormEntryController.ANSWER_CONSTRAINT_VIOLATED:
+            case FormEntryController.ANSWER_REQUIRED_BUT_EMPTY: {
+                if (auditEventLogger != null) {
+                    auditEventLogger.exitView();
+                    auditEventLogger.logEvent(AuditEvent.AuditEventType.CONSTRAINT_ERROR, true, clock.getCurrentTime());
+                }
+
+                setSaveRequestState(taskResult, SaveResult.State.CONSTRAINT_ERROR, saveResult.request);
+                break;
+            }
+        }
+    }
+
+    private boolean requiresReasonToSave() {
+        return auditEventLogger != null
+                && auditEventLogger.isEditing()
+                && auditEventLogger.isChangeReasonRequired()
+                && auditEventLogger.isChangesMade();
+    }
+
+    public static class SaveResult {
+
+        private final State state;
+        private final String message;
+        private final SaveRequest request;
+
+        SaveResult(State state, SaveRequest request) {
+            this(state, request, null);
+        }
+
+        SaveResult(State state, SaveRequest request, String message) {
+            this.state = state;
+            this.message = message;
+            this.request = request;
+        }
+
+        public State getState() {
+            return state;
+        }
+
+        public String getMessage() {
+            return message;
+        }
+
+        public enum State {
+            CHANGE_REASON_REQUIRED,
+            SAVING,
+            SAVED,
+            SAVE_ERROR,
+            FINALIZE_ERROR,
+            CONSTRAINT_ERROR
+        }
+    }
+
+    private static class SaveRequest {
+
+        private final boolean formComplete;
+        private final boolean viewExiting;
+        private final String updatedSaveName;
+        private final Uri uri;
+
+        SaveRequest(boolean formComplete, boolean viewExiting, String updatedSaveName, Uri uri) {
```

Comment:
This may seem petty but I find the order of parameters here make it hard to understand what's important. I'd rename `uri` to `instanceContentUri` and put it first since that's what is requested to be saved.

Reply:
No I agree actually. Will change.